src/main/scala/org/broadinstitute/dsde/firecloud/EntityClient.scala

```diff
@@ -238,32 +239,42 @@ class EntityClient (requestContext: RequestContext)(implicit protected val execu
+    tsvType match {
+      case TsvTypes.MEMBERSHIP => importMembershipTSV(pipeline, workspaceNamespace, workspaceName, tsv, entityType)
+      case TsvTypes.ENTITY => importEntityTSV(pipeline, workspaceNamespace, workspaceName, tsv, entityType)
+      case TsvTypes.UPDATE => importUpdateTSV(pipeline, workspaceNamespace, workspaceName, tsv, entityType)
+      case _ => Future(RequestCompleteWithErrorReport(BadRequest, "Invalid TSV type.")) //We should never get to this case
+    }
+  }
+
   /**
    * Determines the TSV type from the first column header and routes it to the correct import function. */
-  def importEntitiesFromTSV(
-    pipeline: WithTransformerConcatenation[HttpRequest, Future[HttpResponse]],
+  def importEntitiesFromTSV(pipeline: WithTransformerConcatenation[HttpRequest, Future[HttpResponse]],
     workspaceNamespace: String, workspaceName: String, tsvString: String): Future[PerRequestMessage] = {
+
+    def stripEntityType(entityTypeString: String): String = {
+      val entityType = entityTypeString.stripSuffix("_id")
+      if (entityType == entityTypeString)
+        throw new FireCloudExceptionWithErrorReport(errorReport = ErrorReport(StatusCodes.BadRequest, "Invalid first column header, entity type should end in _id"))
+      entityType
+    }
+
     withTSVFile(tsvString) { tsv =>
-      //the first column header of the tsv defines:
-      // 1. The type of the tsv import file (membership, entity, or update)
-      // 2. The entity type we're trying to import or update, which is magically verified in the withFoo functions below.
-      tsv.firstColumnHeader.split(":") match {
-        case Array( tsvType, entityHeader ) =>
-          val entityType = entityHeader.stripSuffix("_id")
-          if( entityType == entityHeader ) {
-            Future(RequestCompleteWithErrorReport(BadRequest, "Invalid first column header, entity type should end in _id"))
-          } else {
-            val strippedTsv = backwardsCompatStripIdSuffixes(tsv, entityType)
-            Try(TsvTypes.withName(tsvType)) match {
-              case Success(TsvTypes.MEMBERSHIP) => importMembershipTSV (pipeline, workspaceNamespace, workspaceName, strippedTsv, entityType)
-              case Success(TsvTypes.ENTITY) => importEntityTSV (pipeline, workspaceNamespace, workspaceName, strippedTsv, entityType)
-              case Success(TsvTypes.UPDATE) => importUpdateTSV (pipeline, workspaceNamespace, workspaceName, strippedTsv, entityType)
-              case Failure(err) => Future(RequestCompleteWithErrorReport(BadRequest, err.toString))
-            }
+      val (tsvType, entityType) = tsv.firstColumnHeader.split(":") match {
+        case Array(entityTypeString) => (TsvTypes.ENTITY, stripEntityType(entityTypeString))
+        case Array(tsvTypeString, entityTypeString) => {
+          val tsvType = Try(TsvTypes.withName(tsvTypeString)) match {
+            case Success(t) => t
```

Comment:
I'm not sure if they're a less verbose way to do a Try() match when you're just passing up the results of a success and then throwing an error that uses the error message if it fails.

Reply:
nah i think this is it. you might be able to do `getOrElse(throw exc)` but that seems _way_ crazy